app/models/globus_destination.rb

```diff
@@ -6,36 +6,55 @@ class GlobusDestination < ApplicationRecord
   belongs_to :user
   after_initialize :set_directory
 
-  # A helper to look up the GlobusDestination object using a Globus URL
+  # A helper to look up the GlobusDestination object using a Globus URL.
+  # @param url [String] a Globus URL
+  # @return [GlobusDestination, nil] the GlobusDestination found or nil
   def self.find_with_globus_url(url)
-    uri = URI.parse(url)
-    return unless uri.query
-
-    params = CGI.parse(uri.query)
-    path = params['destination_path'].first
+    path = extract_path(url)
     return unless path
 
     _, sunet_id, directory = path.split('/')
     user = User.find_by(sunet_id:)
     GlobusDestination.find_by(user:, directory:)
   end
 
-  # creates a URL for globus, including the path to the user's destination directory
+  # Extract the path from either destination_path or origin_path depending on
+  # whether origin_id or destination_id has the configured globus-endpoint-id.
+  # This allows users to paste in a Globus viewer URL that they may have been
+  # emailed which flips around the destination_path to the origin_path, and also
+  # will find when they may have flipped the origin/destination in the viewer.
+  # @param url [String] a Globus URL
+  # @return [String, nil] the path value, or nil if not found
+  def self.extract_path(url)
+    uri = URI.parse(url)
+    return unless uri.query
+
+    params = CGI.parse(uri.query)
+    endpoint_param = if params['origin_id']&.first == Settings.globus.endpoint_id
+                       'origin_path'
+                     else
+                       'destination_path'
+                     end
+
+    params[endpoint_param]&.first
+  end
+
+  # Creates a URL for globus, including the path to the user's destination directory
   def url
     "https://app.globus.org/file-manager?&destination_id=#{Settings.globus.endpoint_id}&destination_path=#{destination_path}"
   end
 
-  # directory within globus including user directory in the format /sunet/datetime
+  # Get the directory within globus including user directory in the format /sunet/datetime
```

Comment:
Adjusted for style. If this isn't correct please let me know!

Reply:
🍻